crates/paymaster-execution/src/diagnostics/extractors/avnu.rs

```diff
@@ -0,0 +1,1135 @@
+                metadata.insert(format!("{prefix}_name"), token_info.name.clone().into());
+            }
+        }
+    }
+
+    /// Adds an amount to metadata, both normalized and raw hex.
+    ///
+    /// Always adds:
+    /// - `{key}_hex`: Raw amount as hex string (e.g., "0xde0b6b3a7640000")
+    ///
+    /// When token info is available, also adds:
+    /// - `{key}`: Normalized amount as Float (e.g., 1.5 instead of 1500000000000000000)
+    async fn add_amount(&self, metadata: &mut HashMap<String, DiagnosticValue>, token_address: Felt, raw_amount: u128, key: &str) {
+        // Always add raw hex value
+        metadata.insert(format!("{key}_hex"), format!("0x{:x}", raw_amount).into());
+
+        // Try to normalize with token decimals
+        if let Some(ref token_service) = self.token_service {
+            if let Some(token_info) = token_service.get(token_address).await {
+                let divisor = 10u128.pow(token_info.decimals as u32);
+                let normalized = raw_amount as f64 / divisor as f64;
+                metadata.insert(key.to_string(), normalized.into());
+            }
+        }
+    }
+
+    /// Extracts parameters from a multi_route_swap call.
+    ///
+    /// Calldata layout:
+    /// 0: sell_token_address
+    /// 1-2: sell_token_amount (u256 = low, high)
+    /// 3: buy_token_address
+    /// 4-5: buy_token_amount (u256)
+    /// 6-7: buy_token_min_amount (u256)
+    /// 8: beneficiary
+    /// 9: integrator_fee_amount_bps
+    /// 10: integrator_fee_recipient
+    /// 11+: routes (Array<Route>)
+    async fn extract_multi_route_swap_params(&self, call: &Call) -> HashMap<String, DiagnosticValue> {
+        let mut metadata = HashMap::new();
+        let calldata = &call.calldata;
+
+        if calldata.len() >= 11 {
+            let sell_token = calldata[0];
+            let sell_amount_raw = felt_to_u128(calldata[1]);
+            let buy_token = calldata[3];
+            let buy_amount_raw = felt_to_u128(calldata[4]);
+            let buy_min_amount_raw = felt_to_u128(calldata[6]);
+
+            // Token addresses
+            metadata.insert("sell_token".to_string(), sell_token.into());
+            metadata.insert("buy_token".to_string(), buy_token.into());
+
+            // Other parameters
+            metadata.insert("beneficiary".to_string(), calldata[8].into());
+            metadata.insert("integrator_fee_bps".to_string(), felt_to_u128(calldata[9]).into());
+            metadata.insert("integrator_fee_recipient".to_string(), calldata[10].into());
+
+            // Enrich token info (symbol, name) - once per token
+            self.enrich_token_info(&mut metadata, sell_token, "sell_token").await;
+            self.enrich_token_info(&mut metadata, buy_token, "buy_token").await;
+
+            // Add amounts (normalized + hex)
+            self.add_amount(&mut metadata, sell_token, sell_amount_raw, "sell_amount").await;
```

Comment:
This is strange, when you have the normalize function I think you can inline everything